compaction/compaction_manager.cc

```diff
@@ -399,25 +401,116 @@ compaction_manager::run_with_compaction_disabled(replica::table* t, std::functio
     co_return;
 }
 
+std::string_view compaction_manager::task::to_string(state s) {
+    switch (s) {
+    case state::init: return "init";
+    case state::pending: return "pending";
+    case state::active: return "active";
+    case state::done: return "done";
+    case state::postponed: return "postponed";
+    case state::failed: return "failed";
+    }
+    __builtin_unreachable();
+}
+
+std::ostream& operator<<(std::ostream& os, compaction_manager::task::state s) {
+    return os << compaction_manager::task::to_string(s);
+}
+
 std::ostream& operator<<(std::ostream& os, const compaction_manager::task& task) {
     return os << task.describe();
 }
 
 std::string compaction_manager::task::describe() const {
-    auto* t = compacting_table();
+    auto* t = _compacting_table;
     auto s = t->schema();
     return fmt::format("{} task {} for table {}.{} [{}]", _description, fmt::ptr(this), s->ks_name(), s->cf_name(), fmt::ptr(t));
 }
 
+compaction_manager::task::~task() {
+    switch_state(state::init);
+}
+
+compaction_manager::sstables_task::~sstables_task() {
+    _cm._stats.pending_tasks -= _sstables.size();
+}
+
+future<> compaction_manager::task::run() noexcept {
+    try {
+        _compaction_done = do_run();
+        return compaction_done();
+    } catch (...) {
+        return current_exception_as_future();
+    }
+}
+
+compaction_manager::task::state compaction_manager::task::switch_state(state new_state) {
+    cmlog.debug("{}: switch_state: {} -> {}", *this, _state, new_state);
+    auto old_state = std::exchange(_state, new_state);
+    switch (old_state) {
+    case state::init:
+    case state::done:
+    case state::postponed:
+    case state::failed:
+        break;
+    case state::pending:
+        --_cm._stats.pending_tasks;
+        break;
+    case state::active:
+        --_cm._stats.active_tasks;
+        break;
+    }
+    switch (new_state) {
+    case state::init:
+        break;
+    case state::pending:
+        ++_cm._stats.pending_tasks;
+        break;
+    case state::active:
+        ++_cm._stats.active_tasks;
+        break;
+    case state::done:
+        ++_cm._stats.completed_tasks;
+        break;
+    case state::postponed:
+    case state::failed:
+        break;
+    }
+    return old_state;
+}
+
+void compaction_manager::sstables_task::set_sstables(std::vector<sstables::shared_sstable> new_sstables) {
+    if (!_sstables.empty()) {
+        on_internal_error(cmlog, format("sstables were already set"));
+    }
+    _sstables = std::move(new_sstables);
+    cmlog.debug("{}: set_sstables: {} sstable{}", *this, _sstables.size(), _sstables.size() > 1 ? "s" : "");
+    _cm._stats.pending_tasks += _sstables.size() - (_state == state::pending);
+}
```

Comment:
as we unconditionally remove _sstables.size() on task dtor, I think it's better to just increment pending_tasks by _sstables.size(). I know that once we bump into semaphores, pending_tasks will be off-by-one, but that's ok I think. pending tasks is hard to reason about as it doesn't say anything about the weight of individual operations.

Reply:
> as we unconditionally remove _sstables.size()

This is no longer the case.
See `sstables_task::~sstables_task()`